File: src/01-dog-age-calculator/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

type AgeResult = number | string;

function calculate_dog_age(age: number): AgeResult {
  if (age < 1) {
    return "Error: age should be at least 1.";
  }
  return Math.round(16 * Math.log(age) + 31);
}

function main() {
  console.log("Dog Age Calculator");
  let age = 1;
  let dog_age = calculate_dog_age(age);
  if (typeof dog_age === "number") {
    console.log(`Dog age ${age} in human years is ${dog_age}`);
  }
  if (typeof dog_age === "string") {
    console.log(`${dog_age}`); // print error
  }
}

main();

File: src/01-dog-age-calculator/haskell/main.hs
{-
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
-}

type AgeResult = Either String Int

calculateDogAge :: Int -> AgeResult
calculateDogAge age
    | age < 1   = Left "Error: age should be at least 1."
    | otherwise = Right $ round (16 * log (fromIntegral age) + 31)

main :: IO ()
main = do
    putStrLn "Dog Age Calculator"
    let age = 1
    case calculateDogAge age of
        Right dogAge -> putStrLn $ "Dog age " ++ show age ++ " in human years is " ++ show dogAge
        Left errorMsg -> putStrLn errorMsg
